net/ipsec/ipsec.go

```diff
@@ -0,0 +1,402 @@
+// called XFRM. Unfortunately, docs are barely existing and the complexity of
+// the framework is high. The best resource I found is Chapter 10 in
+// "Linux Kernel Networking: Implementation and Theory" by Rami Rosen.
+//
+// At the high level, we use the ESP protocol in the Transport mode. Each packet
+// is encrypted with "rfc4106(gcm(aes))", with 32 bytes key and 4 bytes salt.
+// For each connection direction different key is used. An IPsec connection
+// between two peers is identified by directional SPIs formed by concatenating
+// `fromPeerShortID` and `toPeerShortID`.
+//
+// To establish IPsec between peer A and B we need to do the following on each
+// peer:
+//
+//      1. Create (inbound) SA which determines how to process (decrypt) an
+//         incoming packet from the remote peer.
+//      2. Create (outbound) SA for encrypting a packet destined to the remote
+//         peer.
+//      3. Create XFRM policy which says what SA to apply for an outgoing packet.
+//      4. Install iptables rules for marking the vxlan-tunneled outbound
+//         traffic.
+//
+// The kernel VXLAN driver does not set a dst port of a tunnel in the ip flow
+// descriptor, thus xfrm policy lookup cannot match a policy which includes
+// the dst port. To work around, we mark outgoing over the tunnel packets with
+// iptables and set the same mark in the policy selector (funnily enough,
+// iptables_mangle module eventually sets the missing dst port in the flow
+// descriptor). The challenge here is to pick such a mark that it would not
+// interfere with other networking applications before OUTPUT'ing a packet. For
+// example, k8s by default uses 1<<14 and 1<<15 marks.
+package ipsec
+
+// TODO(mp)
+// * Document passwd derivation.
+// * Document MTU requirements.
+// * Test NAT-T in tunnel mode.
+// * Blogpost (w/ benchmarks)
+//
+// * Various XFRM related improvements to vishvananda/netlink.
+// * Patch the kernel.
+// * Extend the heartbeats to check whether encryption is properly set.
+// * Rotate keys.
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+	"syscall"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+
+	"github.com/weaveworks/mesh"
+)
+
+type SPI uint32
+
+const (
+	mask = SPI(2)<<(mesh.PeerShortIDBits-1) - 1
+
+	mark    = uint32(0x1) << 17
+	markStr = "0x20000/0x20000"
+
+	table     = "mangle"
+	markChain = "WEAVE-IPSEC-MARK"
+	mainChain = "WEAVE-IPSEC"
+)
+
+// IPSec
+
+type IPSec struct {
+	sync.RWMutex
+	ipt *iptables.IPTables
+	rc  *connRefCount
+}
+
+func New() (*IPSec, error) {
+	ipt, err := iptables.New()
+	if err != nil {
+		return nil, errors.Wrap(err, "iptables new")
+	}
+
+	ipsec := &IPSec{
+		ipt: ipt,
+		rc:  newConnRefCount(),
+	}
+
+	return ipsec, nil
+}
+
+// Protect establishes IPsec between given peers.
+func (ipsec *IPSec) Protect(srcPeer, dstPeer mesh.PeerShortID, srcIP, dstIP net.IP, dstPort int, localKey, remoteKey []byte) (SPI, error) {
```

Comment:
Partially, because we do not do key rotation, so Seq Nr (4 bytes) in ESP might overflow which creates a possibility for a replay attack.
Btw, I have updated the doc which includes the sec properties we provide.